File: 328. Odd Even Linked List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def oddEvenList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        odd = ListNode(-1)
        even = ListNode(-1)
        evenhead = even
        oddhead = odd
        while head:
            odd.next = head
            head = head.next
            odd = odd.next
            odd.next = None
            if head:
                even.next = head
                head = head.next
                even = even.next
                even.next = None
        odd.next = evenhead.next
        return oddhead.next

File: 328. Odd Even Linked List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* oddEvenList(ListNode* head) {
        ListNode* odd = new ListNode(-1);
        ListNode* even = new ListNode(-1);
        ListNode* oddhead = odd;
        ListNode* evenhead = even;
        while (head != NULL)
        {
            odd->next = head;
            head = head->next;
            odd = odd->next;
            odd->next = NULL;
            if (head != NULL)
            {
                even->next = head;
                head = head->next;
                even = even->next;
                even->next = NULL;
            }
        }
        odd->next = evenhead->next;
        return oddhead->next;
    }
};
